lib/chef/application/exit_code.rb

```diff
@@ -0,0 +1,226 @@
+            exit_code
+          when Exception
+            lookup_exit_code_by_exception(exit_code)
+          else
+            default_exit_code
+          end
+        end
+
+        def normalize_exit_code_to_rfc(exit_code)
+          normalized_exit_code = normalize_legacy_exit_code_with_warning(exit_code)
+          if valid_exit_codes.include? normalized_exit_code
+            normalized_exit_code
+          else
+            VALID_RFC_062_EXIT_CODES[:GENERIC_FAILURE]
+          end
+        end
+
+        def lookup_exit_code_by_exception(exception)
+          if sigint_received?(exception)
+            VALID_RFC_062_EXIT_CODES[:SIGINT_RECEIVED]
+          elsif sigterm_received?(exception)
+            VALID_RFC_062_EXIT_CODES[:SIGTERM_RECEIVED]
+          elsif normalization_disabled? || normalization_not_configured?
+            if legacy_exit_code?(exception)
+              # We have lots of "Chef::Application.fatal!('', 2)
+              # This maintains that behavior at initial introduction
+              # and when the RFC exit_status compliance is disabled.
+              VALID_RFC_062_EXIT_CODES[:SIGINT_RECEIVED]
+            else
+              VALID_RFC_062_EXIT_CODES[:GENERIC_FAILURE]
+            end
+          elsif reboot_scheduled?(exception)
+            VALID_RFC_062_EXIT_CODES[:REBOOT_SCHEDULED]
+          elsif reboot_needed?(exception)
+            VALID_RFC_062_EXIT_CODES[:REBOOT_NEEDED]
+          elsif reboot_failed?(exception)
+            VALID_RFC_062_EXIT_CODES[:REBOOT_FAILED]
+          elsif audit_failure?(exception)
+            VALID_RFC_062_EXIT_CODES[:AUDIT_MODE_FAILURE]
+          else
+            VALID_RFC_062_EXIT_CODES[:GENERIC_FAILURE]
+          end
+        end
+
+        def legacy_exit_code?(exception)
+          resolve_exception_array(exception).any? do |e|
+            e.is_a? Chef::Exceptions::DeprecatedExitCode
+          end
+        end
+
+        def reboot_scheduled?(exception)
+          resolve_exception_array(exception).any? do |e|
+            e.is_a? Chef::Exceptions::Reboot
+          end
+        end
+
+        def reboot_needed?(exception)
+          resolve_exception_array(exception).any? do |e|
+            e.is_a? Chef::Exceptions::RebootPending
+          end
+        end
+
+        def reboot_failed?(exception)
+          resolve_exception_array(exception).any? do |e|
+            e.is_a? Chef::Exceptions::RebootFailed
+          end
+        end
+
+        def audit_failure?(exception)
+          resolve_exception_array(exception).any? do |e|
+            e.is_a? Chef::Exceptions::AuditError
+          end
+        end
+
+        def sigint_received?(exception)
+          resolve_exception_array(exception).any? do |e|
+            e.is_a? Chef::Exceptions::SigInt
+          end
+        end
+
+        def sigterm_received?(exception)
+          resolve_exception_array(exception).any? do |e|
+            e.is_a? Chef::Exceptions::SigTerm
+          end
+        end
+
```

Comment:
I think you could get avoid needing each of these blah_received? methods by either expanding `resole_exception_array` or creating a new method: `resolves_to?(exception, type)`